### ISSUE TITLE
Looking for a way to elide bounds checks

### ISSUE THREAD

**Issue by programmatom** [dotnet/coreclr#3210]
Sorry if this is a duplicate -- I'd guess this is a common request, but oddly couldn't find mention of it in the issues list.

I'm trying to do some heavy numeric computation, including use of SIMD (but that's not directly part of the problem). The issue is that the JIT is never able to elide array bounds checks for my code, since it seems the complexity is over the limit, even though the loops look simple to the human eye.

In typical fashion, I have isolated many of the computation-intensive loops into functions (like BLAS) which are parameterized in a fairly generic way. Here is the simplest example from my collection:

```
public static void FloatVectorSet(
    float[] target,
    int offset,
    int count,
    float value)
{
    if ((count < 0) || unchecked((uint)(offset + count) > (uint)target.Length))
    {
        throw new IndexOutOfRangeException();
    }

    int i = 0;

    Vector<float> vectorValue = new Vector<float>(value);
    for (; i <= count - Vector<float>.Count; i += Vector<float>.Count)
    {
        vectorValue.CopyTo(target, i + offset);
    }

    for (; i < count; i++)
    {
        target[i + offset] = value;
    }
}

```
The resulting Release JIT'ed code is this:
```
FloatVectorSet
00007ffe68ee81d0 56                             push rsi
00007ffe68ee81d1 48 83 ec 20                    sub rsp, 0x20

// Prologue test which covers the loop accesses
00007ffe68ee81d5 45 85 c0                       test r8d, r8d // offset < 0
00007ffe68ee81d8 7c 64                          jl 0x68ee823e
00007ffe68ee81da 42 8d 04 02                    lea eax, [rdx+r8]
00007ffe68ee81de 44 8b 49 08                    mov r9d, [rcx+0x8]
00007ffe68ee81e2 41 3b c1                       cmp eax, r9d // array bounds test
00007ffe68ee81e5 77 57                          ja 0x68ee823e

// SIMD init
00007ffe68ee81e7 33 c0                          xor eax, eax // i = 0
00007ffe68ee81e9 0f 28 c3                       movaps xmm0, xmm3
00007ffe68ee81ec 0f c6 c0 00                    shufps xmm0, xmm0, 0x0
00007ffe68ee81f0 45 8d 50 fc                    lea r10d, [r8-0x4] // loop guard test for no iterations
00007ffe68ee81f4 45 85 d2                       test r10d, r10d
00007ffe68ee81f7 7c 20                          jl 0x68ee8219

// Loop body - note that there are TWO bounds checks generated, so the cost here is even worse than in scalar code case
00007ffe68ee81f9 44 8d 1c 10                    lea r11d, [rax+rdx]
00007ffe68ee81fd 45 3b d9                       cmp r11d, r9d // front end bounds check
00007ffe68ee8200 73 5e                          jae 0x68ee8260
00007ffe68ee8202 41 8d 73 03                    lea esi, [r11+0x3]
00007ffe68ee8206 41 3b f1                       cmp esi, r9d // back end bounds check
00007ffe68ee8209 73 5a                          jae 0x68ee8265
00007ffe68ee820b 42 0f 11 44 99 10              movups [rcx+r11*4+0x10], xmm0
00007ffe68ee8211 83 c0 04                       add eax, 0x4
00007ffe68ee8214 41 3b c2                       cmp eax, r10d
00007ffe68ee8217 7e e0                          jle 0x68ee81f9

00007ffe68ee8219 41 3b c0                       cmp eax, r8d // cleanup loop guard check
00007ffe68ee821c 7d 1a                          jge 0x68ee8238

// Scalar clean-up loop
00007ffe68ee821e 44 8d 14 10                    lea r10d, [rax+rdx]
00007ffe68ee8222 45 3b d1                       cmp r10d, r9d // simple element bounds check
00007ffe68ee8225 73 44                          jae 0x68ee826b
00007ffe68ee8227 4d 63 d2                       movsxd r10, r10d
00007ffe68ee822a f3 42 0f 11 5c 91 10           movss [rcx+r10*4+0x10], xmm3
00007ffe68ee8231 ff c0                          inc eax
00007ffe68ee8233 41 3b c0                       cmp eax, r8d
00007ffe68ee8236 7c e6                          jl 0x68ee821e

// function epilogue
00007ffe68ee8238 48 83 c4 20                    add rsp, 0x20
00007ffe68ee823c 5e                             pop rsi
00007ffe68ee823d c3                             ret

// exception generators
00007ffe68ee823e 48 b9 50 ea 7a c5 fe 7f 00 00  mov rcx, 0x7ffec57aea50
00007ffe68ee8248 e8 63 c9 39 5f                 call 0x7ffec8284bb0
00007ffe68ee824d 48 8b f0                       mov rsi, rax
00007ffe68ee8250 48 8b ce                       mov rcx, rsi
00007ffe68ee8253 e8 38 e6 62 5c                 call 0x7ffec5516890
00007ffe68ee8258 48 8b ce                       mov rcx, rsi
00007ffe68ee825b e8 90 b3 4c 5f                 call 0x7ffec83b35f0
00007ffe68ee8260 e8 df 2b 82 5f                 call 0x7ffec870ae44
00007ffe68ee8265 e8 2a 2b 82 5f                 call 0x7ffec870ad94
00007ffe68ee826a cc                             int3
00007ffe68ee826b e8 68 23 82 5f                 call 0x7ffec870a5d8
00007ffe68ee8270 cc                             int3
```

Discussion:
I understand the difficulty in analyzing this loop:
1. First you have to determine a compile-time closed form for the exact array indices. In graduate school I worked on a parallelizing fortran compiler that did such analysis, and it wasn't simple, so I can get it that the JIT isn't going to figure that one out.
2. Even if you can analyze the loop variables, you still need to determine that the explicit bounds check at the beginning of the function dominates it, which doesn't look trivial (I know the JIT can sometimes do this in trivial cases).

Possible solutions that would help me:

1. Make the JIT do this analysis. I'm listing this for completeness, but guess it's a non-starter. Even if you can manage to do the above closed form array indexing analysis, my guess is it's an expensive analysis that would harm JIT throughput and therefore would not be accepted.

2. Add a (potentially unsafe) way of explicitly declaring the array indices to be in bounds. I see there is already a proposal for a concept of "Unsafe.Assume()" at https://github.com/dotnet/coreclr/issues/2725 although the range of accepted expressions is not defined. Something like **Unsafe.Assume(i + Vector<float>.Count - 1 < target.Length)** as a way of expressing this. Another option would be a less subtle **#pragma suppress-array-bounds-check** or some such thing. Totally unsafe code at that point, of course.

3. In the worst case, if I can't get the perf I need, I could switch to unmanaged code and interop-call it. I'd rather not, because:
A. this code isn't actually unsafe, it's just unanalyzable, and C# gives benefits in writing it, such as a nice SIMD abstraction and potential platform portability.
B. there is a possibility that a p-invoke function call costs more. I haven't tried to quantify this cost at this point.
C. deployment is more costly - since now I have to take a dependency on another dll and whatever runtime libraries get dragged in (e.g. managed C++ makes me suddenly need VS2015 Redistributable, which I don't need at this time).

You might ask why I don't use "fixed" -- because I don't want to incur the cost of creating a pinning GC handle. These loops aren't all that long. There are places I've tried it and it does hurt. In my case it so happens that the base array is already pinned and sub-workspaces are carved out to have Vector<float>-alignment, but I have no way of expressing this to C#, so I would have to re-incur the cost of creating the pinning GC handle for fixed on every call of my function.

So to summarize, I'm either asking for a feature or looking for guidance on how to approach this problem.

**Comment by jkotas** [comment]
> You might ask why I don't use "fixed" -- because I don't want to incur the cost of creating a pinning GC handle

`fixed` does not actually create a pinning GC handle. It just marks a local variable as pinned. The runtime cost - if the array is pinned already - is very low (a few instructions).

**Comment by programmatom** [comment]
I can take a closer look at `fixed` to see what is happening in this case. If it's as cheap as you say, and https://github.com/dotnet/corefx/issues/5106 also happens (adding `Vector<T>.CopyTo(void*)`) it might cover my situation.

**Comment by programmatom** [comment]
OK, I see what you mean - `fixed` is producing efficient code. I can't test with SIMD because of 5106, but with another version (unrolled) I don't see any real problems with it:

```
public unsafe static void FloatVectorSet(
    float[] target,
    int offset,
    int count,
    float value)
{
    if ((count < 0) || unchecked((uint)(offset + count) > (uint)target.Length))
    {
        throw new IndexOutOfRangeException();
    }

    int i = 0;

    fixed (float* pTarget = &(target[offset]))
    {
        for (; i <= count - 4; i += 4)
        {
            pTarget[i + 0] = value;
            pTarget[i + 1] = value;
            pTarget[i + 2] = value;
            pTarget[i + 3] = value;
        }

        for (; i < count; i++)
        {
            pTarget[i] = value;
        }
    }
}
```

```
FloatVectorSet
00007ffe68eb8690 56                             push rsi
00007ffe68eb8691 48 83 ec 30                    sub rsp, 0x30
00007ffe68eb8695 33 c0                          xor eax, eax
00007ffe68eb8697 48 89 44 24 28                 mov [rsp+0x28], rax

// array bounds check
00007ffe68eb869c 45 85 c0                       test r8d, r8d // count < 0
00007ffe68eb869f 0f 8c a2 00 00 00              jl 0x7ffe68eb8747
00007ffe68eb86a5 42 8d 04 02                    lea eax, [rdx+r8] // count + offset
00007ffe68eb86a9 44 8b 49 08                    mov r9d, [rcx+0x8] // target.Length
00007ffe68eb86ad 41 3b c1                       cmp eax, r9d
00007ffe68eb86b0 0f 87 91 00 00 00              ja 0x7ffe68eb8747

00007ffe68eb86b6 33 c0                          xor eax, eax // i = 0

// fixed (float* pTarget = &(target[offset]))
00007ffe68eb86b8 41 3b d1                       cmp edx, r9d // offset >= target.Length
00007ffe68eb86bb 0f 83 a9 00 00 00              jae 0x7ffe68eb876a // throw
00007ffe68eb86c1 48 63 d2                       movsxd rdx, edx
00007ffe68eb86c4 48 8d 4c 91 10                 lea rcx, [rcx+rdx*4+0x10] // &(target[offset])
00007ffe68eb86c9 48 89 4c 24 28                 mov [rsp+0x28], rcx

// unrolled loop guard
00007ffe68eb86ce 41 8d 48 fc                    lea ecx, [r8-0x4]
00007ffe68eb86d2 85 c9                          test ecx, ecx // unrolled loop guard test
00007ffe68eb86d4 7c 4b                          jl 0x68eb8721

// unrolled loop
00007ffe68eb86d6 48 8b 54 24 28                 mov rdx, [rsp+0x28]
00007ffe68eb86db 4c 63 c8                       movsxd r9, eax
00007ffe68eb86de f3 42 0f 11 1c 8a              movss [rdx+r9*4], xmm3
00007ffe68eb86e4 48 8b 54 24 28                 mov rdx, [rsp+0x28]
00007ffe68eb86e9 44 8d 48 01                    lea r9d, [rax+0x1]
00007ffe68eb86ed 4d 63 c9                       movsxd r9, r9d
00007ffe68eb86f0 f3 42 0f 11 1c 8a              movss [rdx+r9*4], xmm3
00007ffe68eb86f6 48 8b 54 24 28                 mov rdx, [rsp+0x28]
00007ffe68eb86fb 44 8d 48 02                    lea r9d, [rax+0x2]
00007ffe68eb86ff 4d 63 c9                       movsxd r9, r9d
00007ffe68eb8702 f3 42 0f 11 1c 8a              movss [rdx+r9*4], xmm3
00007ffe68eb8708 48 8b 54 24 28                 mov rdx, [rsp+0x28]
00007ffe68eb870d 44 8d 48 03                    lea r9d, [rax+0x3]
00007ffe68eb8711 4d 63 c9                       movsxd r9, r9d
00007ffe68eb8714 f3 42 0f 11 1c 8a              movss [rdx+r9*4], xmm3
00007ffe68eb871a 83 c0 04                       add eax, 0x4
00007ffe68eb871d 3b c1                          cmp eax, ecx
00007ffe68eb871f 7e b5                          jle 0x68eb86d6

// cleanup loop guard
00007ffe68eb8721 41 3b c0                       cmp eax, r8d
00007ffe68eb8724 7d 14                          jge 0x68eb873a

// cleanup loop
00007ffe68eb8726 48 8b 4c 24 28                 mov rcx, [rsp+0x28]
00007ffe68eb872b 48 63 d0                       movsxd rdx, eax
00007ffe68eb872e f3 0f 11 1c 91                 movss [rcx+rdx*4], xmm3
00007ffe68eb8733 ff c0                          inc eax
00007ffe68eb8735 41 3b c0                       cmp eax, r8d
00007ffe68eb8738 7c ec                          jl 0x68eb8726

00007ffe68eb873a 33 c9                          xor ecx, ecx
00007ffe68eb873c 48 89 4c 24 28                 mov [rsp+0x28], rcx
00007ffe68eb8741 48 83 c4 30                    add rsp, 0x30
00007ffe68eb8745 5e                             pop rsi
00007ffe68eb8746 c3                             ret

00007ffe68eb8747 48 b9 50 ea 7a c5 fe 7f 00 00  mov rcx, 0x7ffec57aea50
00007ffe68eb8751 e8 5a c4 3c 5f                 call 0x7ffec8284bb0
00007ffe68eb8756 48 8b f0                       mov rsi, rax
00007ffe68eb8759 48 8b ce                       mov rcx, rsi
00007ffe68eb875c e8 2f e1 65 5c                 call 0x7ffec5516890
00007ffe68eb8761 48 8b ce                       mov rcx, rsi
00007ffe68eb8764 e8 87 ae 4f 5f                 call 0x7ffec83b35f0
00007ffe68eb8769 cc                             int3
00007ffe68eb876a e8 69 1e 85 5f                 call 0x7ffec870a5d8
00007ffe68eb876f cc                             int3
```

**Comment by programmatom** [comment]
(also need https://github.com/dotnet/corefx/issues/5474)

**Comment by programmatom** [comment]
no action needed here

**Comment by programmatom** [comment]
For the record, in case someone finds this while searching: I did some benchmarking on one of the loops I'm concerned with. It turns out that using `fixed` to eliminate bounds checks made no difference in timing on either a low end processor (N3520) or somewhat more capable (AMD A10-8700P). Any difference was completely lost in the run-to-run variance, and the minimum run times converged on being the same in at least 3 significant digits.

My guess is that with processor decoders and pipelines today able to decode and issue on the order of 5 instructions per clock and reorder in a 30-instruction window, there's so much parallelism available that the bounds checks slide right through "for free" amongst the actual work being done.

**Comment by briansull** [comment]
Yes, often times the cost of range checks are subsumed by the rest of the operations that take place in the loop.  Only for very tight loops or loops or loops that have high register pressure see measurable costs for array bounds checks.
